### PR TITLE
Apply a series of minor performance optimizations

### PR DESCRIPTION
**Pull request by thiemowmde**

### CODE REVIEW

**Comment by mariushoch** [comment]
Seems this is not documented anywhere, but:
* PHP 7.1.16
```
php > $foo = new DOMNodeList();
php > echo isset($foo[0]);
php >
```

* PHP 5.5.9-1ubuntu4.24
```
php > $foo = new DOMNodeList();
php > echo isset($foo[0]);
PHP Fatal error:  Cannot use object of type DOMNodeList as array in php shell code on line 1
PHP Stack trace:
PHP   1. {main}() php shell code:0
```

**Comment by thiemowmde** [comment]
Uh. This array access was in code I did not touched. Thanks for the super helpful insight! Solved now.

File: src/TagNameMatcher.php [hunk 1]
```diff
@@ -39,9 +39,13 @@ public function describeTo( Description $description ) {
 	 * @return bool
 	 */
 	protected function matchesSafelyWithDiagnosticDescription( $item, Description $mismatchDescription ) {
+		if ( $this->tagNameMatcher->matches( $item->tagName ) ) {
+			return true;
```
Comment (mariushoch) [inline src/TagNameMatcher.php]: This seems to change behavior (but probably for good)… was this on purpose?

File: src/ComplexTagMatcher.php [hunk 2]
```diff
@@ -150,18 +151,14 @@ private function parseHtml( $html ) {
...
-
-		$body = array_shift( $directChildren );
-		$directChildren = iterator_to_array( $body->childNodes );
+		$directChildren = $document->documentElement->childNodes->item( 0 )->childNodes;
 
-		if ( count( $directChildren ) !== 1 ) {
+		if ( $directChildren->length !== 1 ) {
 			throw new InvalidArgumentException(
-				'Expected exactly 1 tag description, got ' . count( $directChildren )
-			);
```
Comment (mariushoch) [inline src/ComplexTagMatcher.php]: doh, sorry… I screwed up here when rebasing :S

File: src/TagNameMatcher.php (hunk 1 above)
Comment (thiemowmde) [reply src/TagNameMatcher.php]: What behavior do you mean? The idea is: If the match succeeds, it is pointless to "describe" a mismatch that never happened.

Comment (mariushoch) [reply src/TagNameMatcher.php]: Thought so… just wanted to make sure you changed that on purpose :)

**Pull request merged**